documentation/book/proc-upgrading-brokers-newer-kafka.adoc

```diff
@@ -0,0 +1,112 @@
+* A `Kafka` resource to be upgraded.
+* You have checked that your `Kafka.spec.kafka.config` contains no options that are not supported in the version of Kafka that you are upgrading to.
+
+.Procedure
+
+. Consult the table below and determine whether the new Kafka version has a different log message format version than the old version. 
++
+include::snip-kafka-versions.adoc[leveloffset=+1]
++
+If the log message format versions are the same proceed to the next step. 
+Otherwise ensure the `Kafka.spec.kafka.config` has the `log.message.format.version` configured to the default for the old version.
++
+For example, if upgrading from 2.0.0:
++
+[source,yaml]
+----
+apiVersion: v1alpha1
+kind: Kafka
+spec:
+  # ...
+  kafka:
+    version: 2.0.0
+    config:
+      log.message.format.version: "2.0"
+      # ...
+----
++
+NOTE: You must format the value of `log.message.format.version` as a string to prevent it from being interpreted as a number.
++
+If `log.message.format.version` is unset then set it and wait for the resulting rolling restart of the Kafka cluster to complete.
++
+. Change the `Kafka.spec.kafka.version` to specify the new version, but leave the `log.message.format.version` as the old version.
+If the image to be used is different from the image for the given version of Kafka configured in the Cluster Operator's `STRIMZI_KAFKA_IMAGES` then configure the `Kafka.spec.kafka.image` as well.
++
+For example, if upgrading from Kafka 2.0.0 to 2.1.0:
++
+[source,yaml]
+----
+apiVersion: v1alpha1
+kind: Kafka
+spec:
+  # ...
+  kafka:
+    version: 2.1.0 <1>
+    config:
+      log.message.format.version: "2.0" <2>
+      # ...
+----
+<1> This is changed to the new version
+<2> This remains at the old version
+
+. Wait for the Cluster Operator to upgrade the cluster.
+If the old and new versions of Kafka have different interbroker protocol versions, look in the Cluster Operator logs for an `INFO` level message like:
++
+[source]
+----
+Reconciliation #_<num>_(watch) Kafka(_<namespace>_/_<name>_): Kafka version upgrade from _<from-version>_ to _<to-version>_, phase 2 of 2 completed
+----
+Alternatively, if the old and new versions of Kafka have the same interbroker protocol version, look in the Cluster Operator logs for an `INFO` level message like:
++
+[source]
+----
+Reconciliation #_<num>_(watch) Kafka(_<namespace>_/_<name>_): Kafka version upgrade from _<from-version>_ to _<to-version>_, phase 1 of 1 completed
+----
++
+For example, using `grep`:
++
+[source,shell,subs="+quotes"]
+----
+oc logs -f _<cluster-operator-pod-name>_ | grep -E "Kafka version upgrade from [0-9.]+ to [0-9.]+, phase ([0-9]+) of \1 completed"
+----
+
+. Upgrade all your client applications to use the new version of the client libraries.
++
+WARNING: You cannot downgrade after completing this step. If, for whatever reason, you need to revert the update at this point, follow the procedure xref:proc-downgrading-brokers-older-kafka-{context}[]. 
```

Comment:
Should this {context} actually be the text of context on the page in question?

Reply:
I don't know. I long ago gave up trying to understand the crazy rules about linking in modular docs. If I replace `{context}` then I'm assuming that the other module has been included in the doc with that context. If I don't replace `{context}` then I'm assuming it will be included with whatever my context is. It seems to me that neither assumption is necessarily justified if we're assuming that modules can be reused arbitrarily (in fact, in that case we cannot even assume that the place we're linking to even exists). On the other hand either would work in in the current context where we're trying to write a single book. So just tell me what you want.

Reply:
Whenever it's an xref the context should be spelled out - you're linking to another section in the guide and the context may not be the same as the module you're putting the xref in.